python/cuml/preprocessing/model_selection.py

```diff
@@ -24,6 +24,109 @@
 from typing import Union
 
 
+def stratify_split(X, y, n_train, n_test, x_numba, y_numba, random_state):
+    """
+    Function to perform a stratified split based on y.
+    Identifies number of classes and samples per class, splices data within
+    each class.
+
+    Input:
+    X, y: shuffled input data and labels
+    """
+    if isinstance(X, cudf.DataFrame):
+        X = cp.fromDlpack(X.to_dlpack())
+    if isinstance(y, cudf.Series):
+        y = cp.fromDlpack(y.to_dlpack())
+    if isinstance(X, cuda.cudadrv.devicearray.DeviceNDArray):
+        X = cp.asarray(X)
+    if isinstance(y, cuda.cudadrv.devicearray.DeviceNDArray):
+        y = cp.asarray(y)
+
+    classes, y_indices = cp.unique(y, return_inverse=True)
+    n_classes = classes.shape[0]
+    class_counts = cp.bincount(y_indices)
+    if n_train < n_classes:
+        raise ValueError('The train_size = %d should be greater or '
+                         'equal to the number of classes = %d' %
+                         (n_train, n_classes))
+    if n_test < n_classes:
+        raise ValueError('The test_size = %d should be greater or '
+                         'equal to the number of classes = %d' %
+                         (n_test, n_classes))
+    class_indices = cp.split(cp.argsort(y_indices),
+                             n_classes)
+
+    X_train = None
+    # Break ties
+    n_i = _approximate_mode(class_counts, n_train, random_state)
+    class_counts_remaining = class_counts - n_i
+    t_i = _approximate_mode(class_counts_remaining, n_test, random_state)
+
+    if random_state is None:
+        rng = np.random.mtrand._rand
```

Comment:
Could we use CuPy or at least NumPy's new Generator (As opposed to the legacy mtrand?)